### PR TITLE
feat(workspaces): change sorting order of the workspace list

### PR DESCRIPTION
**Pull request by rodrimaia**
changes to the sorting order of the workspace list. The new sorting order is as follows:

1. First, display workspaces that are currently running.
2. Next, sort by user_name.
3. Finally, sort by last_used_at in descending order.

That order was suggested in this conversation: https://github.com/coder/coder/pull/7590#pullrequestreview-1432741949.

### CODE REVIEW

**Review by mtojek — COMMENTED** [review]
Let's have a second pair of backend eyes on this PR so that we can be sure that we do not need to mangle with database queries instead.

File: coderd/workspaces.go [hunk 1]
```diff
@@ -1150,6 +1150,19 @@ func convertWorkspaces(workspaces []database.Workspace, data workspaceData) ([]c
 	sort.Slice(apiWorkspaces, func(i, j int) bool {
 		iw := apiWorkspaces[i]
 		jw := apiWorkspaces[j]
+
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply coderd/workspaces.go:1153]: +1 for extracting to its own function, you can then do a nice table-driven unit test

[the remark this answers is not in the document]
Comment (rodrimaia) [reply coderd/workspaces.go:1160]: Great, implementing that! thank you both!  :)

**Review by johnstcn — APPROVED** [review]

File: coderd/workspaces_test.go [hunk 2]
```diff
@@ -198,6 +198,50 @@ func TestAdminViewAllWorkspaces(t *testing.T) {
 	require.Equal(t, 0, len(memberViewWorkspaces.Workspaces), "member in other org should see 0 workspaces")
 }
 
+func TestWorkspacesSortOrder(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (rodrimaia) [reply coderd/workspaces_test.go]: done! :)

File: coderd/workspaces_internal_test.go [hunk 3]
```diff
@@ -80,3 +82,106 @@ func Test_calculateDeletingAt(t *testing.T) {
...
+
+		// in this case, the last used at is the creation time
+		require.Equal(t, workspaces[0].Name, "test-workspace-sort-4")
+		require.Equal(t, workspaces[1].Name, "test-workspace-sort-3")
+		require.Equal(t, workspaces[2].Name, "test-workspace-sort-2")
+		require.Equal(t, workspaces[3].Name, "test-workspace-sort-1")
+	})
+}
+
+func WorkspaceFactory(t *testing.T, name string, ownerID uuid.UUID, ownerName string, status codersdk.WorkspaceStatus) codersdk.Workspace {
```
[the remark this answers is not in the document]
Comment (rodrimaia) [reply coderd/workspaces_internal_test.go]: oh, good to know. thank you

**Comment by rodrimaia** [comment]
@mtojek @johnstcn using table tests now

**Review by mtojek — APPROVED** [review]
👍

**Review by mtojek — COMMENTED** [review]
LGTM, but CI complains about some tests:

```=== Failed
=== FAIL: coderd TestSortWorkspaces/then_sort_by_last_used_at_(recent_first) (0.00s)
    workspaces_internal_test.go:173: 
        	Error Trace:	C:/a/coder/coder/coderd/workspaces_internal_test.go:173
        	Error:      	Not equal: 
        	            	expected: []string{"running-userA3", "running-userA2", "running-userA"}
        	            	actual  : []string{"running-userA3", "running-userA", "running-userA2"}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -2,4 +2,4 @@
        	            	  (string) (len=14) "running-userA3",
        	            	- (string) (len=14) "running-userA2",
        	            	- (string) (len=13) "running-userA"
        	            	+ (string) (len=13) "running-userA",
        	            	+ (string) (len=14) "running-userA2"
        	            	 }
        	Test:       	TestSortWorkspaces/then_sort_by_last_used_at_(recent_first)
        	Messages:   	then sort by last used at (recent first)
```

**Pull request merged**